tests/importer/model_flattening.cpp

```diff
@@ -832,3 +830,132 @@ TEST(ModelFlattening, importedComponentsWithConnectionsToChildren)
+
+    // Resolve the imports.
+    importer->resolveImports(originalModel, resourcePath("modelflattening/"));
+
+    // Check for issues: expect one reporting the circular import.
+    EXPECT_EQ(size_t(1), importer->issueCount());
+    EXPECT_EQ(resolveError, importer->issue(0)->description());
+
+    // Create a flattened version to demonstrate the diagnostics.
+    importer->removeAllIssues();
+    auto flatModel = importer->flattenModel(originalModel);
+    EXPECT_EQ(size_t(1), importer->issueCount());
+    EXPECT_EQ(flattenError, importer->issue(0)->description());
+    EXPECT_EQ(nullptr, flatModel);
+}
+
+TEST(ModelFlattening, resolveFlattenCircularImportsUnits)
+{
+    std::string resolveError = "Cyclic dependencies were found when attempting to resolve units in model 'importExampleUnits'. The dependency loop is:\n"
+                               " - units 'sideB' is imported from 'myChildIsCircular' in 'circularImport1units.cellml';\n"
+                               " - units 'shared' is imported from 'circular2' in 'circularImport2units.cellml';\n"
+                               " - units 'circular2' is imported from 'shared' in 'circularImport1units.cellml'; and\n"
+                               " - units 'shared' is imported from 'circular2' in 'circularImport2units.cellml'.";
+    std::string flattenError = "Cyclic dependencies were found when attempting to flatten units in model 'importExampleUnits'. The dependency loop is:\n"
+                               " - units 'sideB' is imported from 'myChildIsCircular' in 'circularImport1units.cellml';\n"
+                               " - units 'shared' is imported from 'circular2' in 'circularImport2units.cellml';\n"
+                               " - units 'circular2' is imported from 'shared' in 'circularImport1units.cellml'; and\n"
+                               " - units 'shared' is imported from 'circular2' in 'circularImport2units.cellml'.";
+
+    auto parser = libcellml::Parser::create();
+    auto originalModel = parser->parseModel(fileContents("modelflattening/importExampleUnits.cellml"));
+    auto importer = libcellml::Importer::create();
+
+    // Resolve the imports.
+    importer->resolveImports(originalModel, resourcePath("modelflattening/"));
+
+    // Check for issues: expect one reporting the circular import.
+    EXPECT_EQ(size_t(1), importer->issueCount());
+    EXPECT_EQ(resolveError, importer->issue(0)->description());
+
+    // Create a flattened version to demonstrate the diagnostics.
+    importer->removeAllIssues();
+    auto flatModel = importer->flattenModel(originalModel);
+    EXPECT_EQ(size_t(1), importer->issueCount());
+    EXPECT_EQ(flattenError, importer->issue(0)->description());
+    EXPECT_EQ(nullptr, flatModel);
+}
+
+TEST(ModelFlattening, resolveFlattenMissingModel)
+{
+    auto e = "Component 'left' requires a model imported from 'diamond_point.cellml' which is not available in the importer.";
```

Comment:
Besides `auto` that probably ought to be `std::string`, it should also be made `const` and `e` probably ought to be renamed to `error` (all that for consistency).

Reply:
Same elsewhere in this PR.

Reply:
1960e45a